### PR TITLE
Check gradient with OpInfo tests

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Compare inductor's computed gradient with eager computed gradient with OpInfo samples.

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -243,6 +244,43 @@ def run(*ex, **kwargs):
...
+                if arg.requires_grad:
+                    leaf_tensors.append(arg)
+            return leaf_tensors
+
+        def compute_grads(args, kwrags, results):
+            flat_results, _ = tree_flatten(results)
+            flat_diff_results = [r for r in flat_results if r.requires_grad]
+            assert len(flat_diff_results) > 0
+
+            grads = [torch.ones(r.shape, device=r.device, dtype=r.dtype) for r in flat_diff_results]
```
Comment (ngimel) [inline test/inductor/test_torchinductor.py:265]: better to have random grads, `torch.ones` can hide the bugs

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -243,6 +244,43 @@ def run(*ex, **kwargs):
...
+
+            leaf_tensors = gather_leaf_tensors(ref_inputs, ref_kwargs)
+            assert len(leaf_tensors) > 0
+            return torch.autograd.grad(flat_diff_results, leaf_tensors,
+                                    grads, allow_unused=True, retain_graph=True)
+
+        correct_grad = compute_grads(ref_inputs, ref_kwargs, correct)
```
Comment (ngimel) [inline test/inductor/test_torchinductor.py:272]: should we be using gradcheck as @albanD suggested?

Comment (SherlockNoMad) [reply test/inductor/test_torchinductor.py:281]: IIUC, autograd.gradcheck computes gradients via small finite differences. It's useful for checking gradient's correctness when a reference implementation is not available. 

For inductor, we can use eager's result as the reference.

File: test/inductor/test_torchinductor.py [hunk 3]
```diff
@@ -243,6 +244,52 @@ def run(*ex, **kwargs):
...
+            leaf_tensors = []
+            args, args_spec = tree_flatten(args)
+            kwargs, kwargs_spec = tree_flatten(kwargs)
+            args = args + kwargs
+            for arg in args:
+                if not isinstance(arg, torch.Tensor):
+                    continue
+                if arg.requires_grad:
+                    leaf_tensors.append(arg)
+            return leaf_tensors
```
Comment (voznesenskym) [inline test/inductor/test_torchinductor.py:259]: spec is unused, and args and kwargs are both flattened - can we do something like:

`leaf_tensors = [arg for arg in tree_flatten(args, kwargs)[0] if isinstance(arg, torch.Tensor)]`

(Note: did not run the above, but list comprehension should be able to one line it)

File: test/inductor/test_torchinductor.py [hunk 4]
```diff
@@ -243,6 +244,52 @@ def run(*ex, **kwargs):
...
+            assert len(flat_diff_results) > 0
+
+            grads = [
+                torch.ones(r.shape, device=r.device, dtype=r.dtype)
+                for r in flat_diff_results
+            ]
+
+            leaf_tensors = gather_leaf_tensors(ref_inputs, ref_kwargs)
+            assert len(leaf_tensors) > 0
+            return torch.autograd.grad(
```
Comment (voznesenskym) [inline test/inductor/test_torchinductor.py:273]: should len of leaf_tensors match len of grads?

File: test/inductor/test_torchinductor.py [hunk 5]
```diff
@@ -243,6 +244,52 @@ def run(*ex, **kwargs):
...
+            kwargs, kwargs_spec = tree_flatten(kwargs)
+            args = args + kwargs
+            for arg in args:
+                if not isinstance(arg, torch.Tensor):
+                    continue
+                if arg.requires_grad:
+                    leaf_tensors.append(arg)
+            return leaf_tensors
+
+        def compute_grads(args, kwrags, results):
```
Comment (voznesenskym) [inline test/inductor/test_torchinductor.py:261]: Nit: maybe pull to top level? It took me a second to realize that it wasn't calling itself in the compute_grads below.

File: test/inductor/test_torchinductor_opinfo.py [hunk 6]
```diff
@@ -394,6 +403,11 @@ def wrapper_set_seed(op, *args, **kwargs):
     "new_empty_strided": {"assert_equal": False},
     "randn": {"assert_equal": False},
     ("nn.functional.tanhshrink", "cuda", f16): {"atol": 3e-4, "rtol": 0.001},
+    "gradient": {"check_gradient": False},  # segfault on check_gradient
+    "linalg.solve_triangular": {"check_gradient": False},  # segfault on check_gradient
+    "linalg.lu_factor": {"check_gradient": False},
+    "linalg.lu_factor_ex": {"check_gradient": False},
+    "linalg.lu_solve": {"check_gradient": False},
```
Comment (voznesenskym) [inline test/inductor/test_torchinductor_opinfo.py:410]: Can you comment why these are skipped as well?

**Review by voznesenskym — APPROVED** [review]
LGTM - comments for code cleanliness and brevity.

File: test/inductor/test_torchinductor.py (hunk 3 above)
Comment (SherlockNoMad) [reply test/inductor/test_torchinductor.py:259]: update.

File: test/inductor/test_torchinductor.py (hunk 4 above)
Comment (SherlockNoMad) [reply test/inductor/test_torchinductor.py:273]: no... 
leaf_tensors are the inputs that requires grad
`grads` are the outputs that requires grad

File: test/inductor/test_torchinductor_opinfo.py [hunk 7]
```diff
@@ -443,18 +458,22 @@ def test_comprehensive(self, device, dtype, op):
             self.skipTest(f"{op_name} in {dtype} not supported")
         elif dtype in inductor_expected_failures_single_sample[device_type].get(
             op_name, set()
+        ) or dtype in inductor_gradient_expected_failures_single_sample[
```
Comment (voznesenskym) [inline test/inductor/test_torchinductor_opinfo.py:461]: If there's a large list of grad failures, we should consider splitting the test into forward or exclude, and then backward or exclude (instead of forward+backward or exclude from either). So as not to regress forward tests.

File: test/inductor/test_torchinductor.py (hunk 5 above)
Comment (SherlockNoMad) [reply test/inductor/test_torchinductor.py:261]: ok.

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (albanD) [reply test/inductor/test_torchinductor.py:277]: The benefit of using gradcheck is that you don't have to write all the code to get proper grad outputs and do the comparison by hand.

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (albanD) [reply test/inductor/test_torchinductor.py]: This might be numerically unstable for big Tensor, for gradcheck, we use a unit-norm Tensor as it proved to be the most stable.

File: test/inductor/test_torchinductor_opinfo.py [hunk 8]
```diff
@@ -394,6 +430,11 @@ def wrapper_set_seed(op, *args, **kwargs):
     "new_empty_strided": {"assert_equal": False},
     "randn": {"assert_equal": False},
     ("nn.functional.tanhshrink", "cuda", f16): {"atol": 3e-4, "rtol": 0.001},
+    "gradient": {"check_gradient": False},  # segfault on check_gradient
+    "linalg.solve_triangular": {"check_gradient": False},
+    "linalg.lu_factor": {"check_gradient": False},
+    # "linalg.lu_factor_ex": {"check_gradient": False},
```
Comment (albanD) [inline test/inductor/test_torchinductor_opinfo.py:436]: remove these?

File: test/inductor/test_torchinductor_opinfo.py [hunk 9]
```diff
@@ -489,24 +534,35 @@ def fn(*args, **kwargs):
...
-                        copy_to_cuda=False,
-                        reference_in_float=False,
-                        **additional_kwargs,
+                        **adjusted_kwargs,
                     )
                 elif device_type == "cpu":
+                    adjusted_kwargs = {
+                        "check_lowp": False,
+                        "nopython": True,
+                        # skip checking gradient on CPU for now
```
Comment (albanD) [inline test/inductor/test_torchinductor_opinfo.py:556]: Why?

File: test/inductor/test_torchinductor_opinfo.py (hunk 6 above)
Comment (SherlockNoMad) [reply test/inductor/test_torchinductor_opinfo.py]: commented.

File: test/inductor/test_torchinductor_opinfo.py (hunk 7 above)
Comment (SherlockNoMad) [reply test/inductor/test_torchinductor_opinfo.py:488]: We have 31 new ops failing due to gradient check... which is manageable...

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (SherlockNoMad) [reply test/inductor/test_torchinductor.py:277]: As discussed offline with Alban, I will stick with current approach, coz gradcheck doesn't work well with low precision floats.

Comment (ngimel) [reply test/inductor/test_torchinductor.py:277]: Good point about fp32 (not really low precision :-) )

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (SherlockNoMad) [reply test/inductor/test_torchinductor.py]: Updated to use unit norm now.

File: test/inductor/test_torchinductor_opinfo.py (hunk 9 above)
Comment (SherlockNoMad) [reply test/inductor/test_torchinductor_opinfo.py:561]: Inductor with CPU is not the focus now... We can enable this when CPU becomes a priority.
